Merge initial FreeRTOS Code into v2

codecubepi opened the pull request:
## PR Overview

This pull request includes my work on the FreeRTOS portion of the `v2` firmware intiative up to 08 Feb 2024.

This includes the FreeRTOS source code v10.6.2 linked as a Git submodule, dual-core applications (without Inter-Core Communication), and a restructure of the `sdk` directory tree such that common system-level drivers can exist in a "shared" folder, which is linked into both dual-core applications (`freertos_app_cpu0` and `freertos_app_cpu1`).

Additionally, efforts have been made in this pull request to streamline the setup of SDK projects for new users by making all build paths in the `.cproject` workspace-relative instead of absolute.

## Dual-Core Apps

The dual-core applications `freertos_app_cpu0` and `freertos_app_cpu1` are to run on _CPU0_ and _CPU1_ respectively to demo basic FreeRTOS task and structure functionality. Both apps run independently of each other. Each app has its own Tx task, Rx task, and a Queue to send messages through. CPU0 operates the RGB LEDs, while messages are transmitting, they cycle red, and when messages have completed, all LEDs blink green (to showcase operation of the `led.c` driver). Each CPU also has its own `FreeRTOSConfig.h`, which are mostly identical, but can be individually configured if needed.

Using a live submodule of the FreeRTOS Kernel will ensure going forward that we can make use of the latest features in FreeRTOS, and updating the kernel is as easy as updating the submodule. 

## Merge Plan

The plan for merging is to **Squash and Merge** the stable FreeRTOS code in this branch into `v2-staging`. A **Squash and Merge** is ideal due to the existence of files created by a commit in this branch, and that are then deleted by a subsequent commit.

codecubepi committed:
Create basic FreeRTOS apps

codecubepi committed:
Update linker script to fix dual-core freezing

codecubepi committed:
Rename hello_world starter code to main.c

codecubepi committed:
Include basic blinky code

codecubepi committed:
Create Inter-core communication files

codecubepi committed:
Add FreeRTOS Kernel as Git submodule

codecubepi committed:
Create FreeRTOSConfig.h for FreeRTOS v10.1.1

codecubepi committed:
Add configUSE_SB_COMPLETED_CALLBACK to freertos_app_cpu0/src/FreeRTOSConfig.h

codecubepi committed:
Update freertos_app_cpu0 files for first draft of ICC

codecubepi committed:
Commit versions of app0 that build, and app1 old files for reference

codecubepi committed:
Update FreeRTOS app0 and app1, both in building state

codecubepi committed:
Convert include paths to workspace-relative paths for both FreeRTOS apps

codecubepi committed:
Add pointer to FreeRTOS-Kernel commit 18ed888 (v10.6.2) in .gitmodules

codecubepi committed:
Update linker script and main() to include FreeRTOS asm vectors

codecubepi committed:
Create clang-format scripts that target FreeRTOS apps

codecubepi committed:
Apply clang-format to FreeRTOS apps

codecubepi committed:
Move sys and drv code to new shared folder, update sdk project refs

codecubepi committed:
Delete outdated README

codecubepi committed:
Resolve warning in cpu1 main.c

codecubepi committed:
Update README in sdk folder

codecubepi committed:
Add missing links in sdk/README

codecubepi commented:
@npetersen2 

Please review this PR when you have time to do so. Tagging #360 for relevance.

codecubepi committed:
Bring updates from "unstable" FreeRTOS branch into stable branch

npetersen2 commented on sdk/FreeRTOS-Kernel:
Hmm.. the commit shown in GitHub doesn't seem to match the commit you have in the .gitmodules. odd

codecubepi replied:
Yes, I think I ran into this issue last semester. I ended up having to manually check out the commit I wanted. I will leave this unresolved so we can discuss in our meeting tomorrow (Tuesday, 13 Feb 2024)

npetersen2 replied:
Let's keep this on the back burner...

npetersen2 commented on sdk/README.md:
```diff
@@ -17,8 +19,31 @@ Combination of both `app_cpu0` and `app_cpu1`:
+
+- Designed for AMDC hardware platform (using PicoZed processor module)
+- Dual-core AMP design
+- Based on [FreeRTOS]()
```
link

codecubepi replied:
Good catch, thanks.

npetersen2 commented on sdk/freertos_app_cpu0/.project:
```diff
@@ -0,0 +1,38 @@
+		<link>
+			<name>FreeRTOS-Kernel</name>
+			<type>2</type>
+			<location>S:/School/WEMPEC/firmware-rtos-e/sdk/FreeRTOS-Kernel</location>
```
Absolute links (probably don't need to fix now, but eventually)

codecubepi replied:
Hmm... not sure what to do about these. As far as I can tell, I've set all of the references to be workspace-relative paths in the properties menu for each sdk project (and they appear that way in the `.cproject` file.) I could try manually editing these absolute paths in these `.project` files in each `freertos-app-cpuX`, but I'm not sure whether or not that will cause issues. 

I'll leave these unresolved for now.

npetersen2 replied:
Let's keep this on the back burner

npetersen2 commented on sdk/freertos_app_cpu0/.project:
```diff
@@ -0,0 +1,38 @@
+		<link>
+			<name>shared</name>
+			<type>2</type>
+			<location>S:/School/WEMPEC/firmware-rtos-e/sdk/shared</location>
```
Absolute links (probably don't need to fix now, but eventually)

npetersen2 commented on sdk/freertos_app_cpu0/src/FreeRTOSConfig.h:
```diff
@@ -0,0 +1,245 @@
+/* README!!!
+ *
+ * This configuration is copied from the FreeRTOS v10.1.1 based example project for the Zynq 7000 platform (for
+ * compatibility). However, you can add to it.
+ *
+ * Another example configuration file is available at
+ * "sdk/FreeRTOS-Kernel/examples/sample_configuration/FreeRTOSConfig.h" Because that example is in the submodule, it may
+ * update from time-to-time when the submodule is updated.
```
Ensure line width is limited like the rest of this comment block

npetersen2 commented on sdk/freertos_app_cpu0/src/FreeRTOSConfig.h:
```diff
@@ -0,0 +1,245 @@
+ */
+
+/*
+ * Copyright (C) 2018 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
```
Lol "Amazon.com"

codecubepi replied:
Should we/can we remove these? They're kinda ugly but it feels illegal to remove a copyright. I'll leave unresolved.

npetersen2 replied:
Let's leave it

npetersen2 commented on sdk/freertos_app_cpu0/src/FreeRTOSConfig.h:
```diff
@@ -0,0 +1,245 @@
+#define configSUPPORT_DYNAMIC_ALLOCATION        1
+#define configSUPPORT_STATIC_ALLOCATION         1
+
+/* WEMPEC ADDITION: Need to specify this callback config so we can create message buffers with callbacks
```
"AMDC" ADDITION

npetersen2 replied:
Shorter comment line length

codecubepi replied:
Nothing to do with the fact that the AMDC is losing its WEMPEC exclusivity, huh? 😉

npetersen2 commented on sdk/freertos_app_cpu0/src/Xilinx.spec:
I'm not sure this file needs to be in git... unknown. But, probably shouldn't spend too much time on this now. I remember I dumped many hours trying to figure out the purpose of this file when I was setting up the `v1` code base... but cannot remember now. Probably should just do what I did back then

codecubepi replied:
Will leave unresolved and discuss in meeting. Since we're squashing when we merge, it is not too late to do a delete commit and erase this from the `v2` development history if that's what you want.

npetersen2 replied:
This `.spec` file is in the v1 codebase on git. Let's keep it.

Looks like a file that goes into the `gcc` compiler commands: https://gcc.gnu.org/onlinedocs/gcc/Spec-Files.html

npetersen2 commented on sdk/freertos_app_cpu0/src/lscript.ld:
```diff
@@ -0,0 +1,289 @@
+
+MEMORY
+{
+   ps7_ddr_0 : ORIGIN = 0x100000, LENGTH = 0x1FF80000
+   ps7_qspi_linear_0 : ORIGIN = 0xFC000000, LENGTH = 0x1000000
+   ps7_ram_0 : ORIGIN = 0x0, LENGTH = 0x30000
+   ps7_ram_1 : ORIGIN = 0xFFFF0000, LENGTH = 0xFE00
```
It says at the top of this file it was auto-generated... but didn't we change some things here?

codecubepi replied:
Yes we did, see caff3f7

Should I remove the "auto-generated" message?

npetersen2 replied:
Let's update the header comment to explain the story: this was originally auto gen, but we modified it per the docs found here: https://docs.amdc.dev/firmware/xilinx-tools/dual-core.html#configure-dram-memory-regions

npetersen2 commented on sdk/freertos_app_cpu0/src/main.c:
I did not review this file, since it will probably change a lot later

npetersen2 commented on sdk/freertos_app_cpu1/.project:
```diff
@@ -0,0 +1,38 @@
+		<link>
+			<name>FreeRTOS-Kernel</name>
+			<type>2</type>
+			<location>S:/School/WEMPEC/firmware-rtos-e/sdk/FreeRTOS-Kernel</location>
+		</link>
+		<link>
+			<name>shared</name>
+			<type>2</type>
+			<location>S:/School/WEMPEC/firmware-rtos-e/sdk/shared</location>
```
Same comment as other core

npetersen2 commented on sdk/freertos_app_cpu1/src/FreeRTOSConfig.h:
```diff
@@ -0,0 +1,245 @@
+/* README!!!
+ *
+ * This configuration is copied from the FreeRTOS v10.1.1 based example project for the Zynq 7000 platform (for
+ * compatibility). However, you can add to it.
+ *
+ * Another example configuration file is available at
+ * "sdk/FreeRTOS-Kernel/examples/sample_configuration/FreeRTOSConfig.h" Because that example is in the submodule, it may
+ * update from time-to-time when the submodule is updated.
```
Comment line length

npetersen2 commented on sdk/freertos_app_cpu1/src/FreeRTOSConfig.h:
```diff
@@ -0,0 +1,245 @@
+#define configSUPPORT_DYNAMIC_ALLOCATION        1
+#define configSUPPORT_STATIC_ALLOCATION         1
+
+/* WEMPEC ADDITION: Need to specify this callback config so we can create message buffers with callbacks
```
AMDC

npetersen2 commented on sdk/freertos_app_cpu1/src/lscript.ld:
```diff
@@ -0,0 +1,289 @@
+/*******************************************************************/
+/*                                                                 */
+/* This file is automatically generated by linker script generator.*/
+/*                                                                 */
+/* Version: 2019.1                                                 */
+/*                                                                 */
+/* Copyright (c) 2010-2016 Xilinx, Inc.  All rights reserved.      */
+/*                                                                 */
+/* Description : Cortex-A9 Linker Script                           */
+/*                                                                 */
+/*******************************************************************/
+
```
Same comment as other core

npetersen2 replied:
Same as other core

npetersen2 commented on sdk/freertos_app_cpu1/src/main.c:
Not reviewing this file now, as it will change a lot

npetersen2 commented on sdk/shared/FreeRTOS_tick_config.c:
```diff
@@ -0,0 +1,138 @@
+#include "xscugic.h"
+#include "xscutimer.h"
+
+#define XSCUTIMER_CLOCK_HZ (XPAR_CPU_CORTEXA9_0_CPU_CLK_FREQ_HZ / 2UL)
```
Huh.. 1/2 global cpu clock freq

npetersen2 commented on sdk/shared/FreeRTOS_tick_config.c:
```diff
@@ -0,0 +1,138 @@
+    interrupt should be left disabled.  It is enabled automatically when the
+    scheduler is started. */
+
+    /* Ensure XScuGic_CfgInitialize() has been called.  In this demo it has
+    already been called from prvSetupHardware() in main(). */
```
Hm I don't see your `prvSetupHardware()` being called in main.c... Does the tick interrupt work for you?

i.e., you can make a task sleep for X seconds and it will wake up correctly? Maybe your blinky does this?

npetersen2 replied:
Seems like your blinky task is using the FreeRTOS sleep methods, and it works... so probably fine

https://github.com/Severson-Group/AMDC-Firmware/blob/feature/freertos-stable/sdk/freertos_app_cpu0/src/main.c#L284-L298

npetersen2 commented on sdk/shared/FreeRTOS_tick_config.c:
```diff
@@ -0,0 +1,138 @@
+}
+/*-----------------------------------------------------------*/
+
+/* This is the callback function which is called by the FreeRTOS Cortex-A port
+layer in response to an interrupt.  If the function is called
+vApplicationFPUSafeIRQHandler() then it is called after the floating point
+registers have been saved.  If the function is called vApplicationIRQHandler()
+then it will be called without first having saved the FPU registers.  See
+http://www.freertos.org/Using-FreeRTOS-on-Cortex-A-Embedded-Processors.html for
+more information */
+void vApplicationFPUSafeIRQHandler(uint32_t ulICCIAR)
+{
+    extern const XScuGic_Config XScuGic_ConfigTable[];
+    static const XScuGic_VectorTableEntry *pxVectorTable
+        = XScuGic_ConfigTable[XPAR_SCUGIC_SINGLE_DEVICE_ID].HandlerTable;
+    uint32_t ulInterruptID;
+    const XScuGic_VectorTableEntry *pxVectorEntry;
+
+    /* Re-enable interrupts. */
+    __asm("cpsie i");
+
+    /* The ID of the interrupt is obtained by bitwise anding the ICCIAR value
+    with 0x3FF. */
+    ulInterruptID = ulICCIAR & 0x3FFUL;
+    if (ulInterruptID < XSCUGIC_MAX_NUM_INTR_INPUTS) {
+        /* Call the function installed in the array of installed handler functions. */
+        pxVectorEntry = &(pxVectorTable[ulInterruptID]);
+        pxVectorEntry->Handler(pxVectorEntry->CallBackRef);
+    }
```
Oh boy haha

codecubepi replied:
Some of this junk that's strewn about may be from the FreeRTOS example projects I started with. The technique is "start with a working example, then remove until you're down to what you want for your app". Unfortunately, there's so much throughout this code that I would have to tinker with deleting to learn what I should/should not delete...

npetersen2 replied:
This is probably needed, so no worries. Just seems quite complex!

npetersen2 commented on sdk/shared/README.md:
```diff
@@ -0,0 +1,30 @@
+# Shared Code Folder
+
+This folder (`AMDC-Firmware/sdk/shared`) was created to contain source code shared by both CPU applications in the v2, FreeRTOS-based firmware.
```
No passive voice.

> This folder () contains source code...

codecubepi replied:
Oof my grammar rightfully called out.

npetersen2 commented on sdk/shared/README.md:
````diff
@@ -0,0 +1,30 @@
+
+Shared code includes system-level source code for inter-core communication, interrupts, etc, as well as the [FreeRTOS-Kernel](https://github.com/FreeRTOS/FreeRTOS-Kernel) source code, which is available in `AMDC-Firmware/sdk/FreeRTOS-Kernel` as a [Git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules).
+
+### Differentiating code between CPUs
+
+If you need to include or run slightly different code on CPU 0 and CPU 1 in a shared source file, it can be done using the `#if` directive like so:
+
+```c
+#include xparameters.h          // Must be included to gain access to the XPAR_CPU_ID definition
+#include foo.h
+#include bar.h
+
+// Code to be included on both CPUs
+
+#if XPAR_CPU_ID == 0
+
+// Code for CPU 0 only
+
+#elif XPAR_CPU_ID == 1
+
+// Code for CPU 1 only
+
+#endif
+
+// More code to be included on both CPUs
+
+```
````
Nice and helpful documentation here! :)

codecubepi replied:
Thanks haha 🙏

npetersen2 commented on sdk/shared/drv/led.c:
Nice, we'll need to correctly merge this with @annikaolson code..... i.e., after merging #358 we will probably move all the driver code to this shared folder, etc.

codecubepi replied:
Sounds good. I will leave this unresolved for now.

npetersen2 commented on sdk/shared/sys/icc.c:
```diff
@@ -0,0 +1,102 @@
+/*
+ * icc.c
+ *
+ *  Created on: 7 nov. 2023
+ *      Author: pnowa
+ */
```
Eh, remove this since it doesn't follow AMDC-Firmware styling

npetersen2 commented on sdk/shared/sys/icc.c:
```diff
@@ -0,0 +1,102 @@
+
+#include "icc.h"
+
+///////////////////////////////////////////////////////
+// KEEP THIS FILE IN SYNC IN BOTH CPU0 AND CPU1
+///////////////////////////////////////////////////////
```
Hm. Maybe this should say something else since this exact file is include for both cores?

codecubepi replied:
Yep, I noticed this as well and corrected it in my work today.

npetersen2 commented on sdk/shared/sys/icc.c:
```diff
@@ -0,0 +1,102 @@
+// KEEP THIS FILE IN SYNC IN BOTH CPU0 AND CPU1
+///////////////////////////////////////////////////////
+
+void icc_init(void)
```
I'll stop reviewing this file since you are working on this now

npetersen2 reviewed (changes requested):
@codecubepi I took a look at all the files changed and left comments.

Mostly ready to be merged, but a few minor things to address

codecubepi committed:
Address PR feedback

codecubepi commented:
@npetersen2 

I have pushed commits that include the work we did in the lab today, as well as a commit addressing most of your feedback. Anything I wasn't able to address I left unresolved for tomorrow's meeting (13 Feb 2024).

codecubepi committed:
Update header comment in linker script

npetersen2 reviewed (approved):
I approve this to be merged, please wait until after #358 is merged.

npetersen2 commented:
We have split the timing manager feature away from the FreeRTOS feature. So, let's merge this PR into a new `v2-staging` and then we will treat `v2-staging` as the stable FreeRTOS codebase.
